extensions/amp-story-auto-ads/0.1/story-ad-config.js

```diff
@@ -0,0 +1,124 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {isJsonScriptTag} from '../../../src/dom';
+import {isObject} from 'util';
+import {parseJson} from '../../../src/json';
+import {startsWith} from '../../../src/string';
+import {user, userAssert} from '../../../src/log';
+
+/** @const {string} */
+const TAG = 'amp-story-auto-ads:config';
+
+/** @const {string} */
+const MUSTACHE_TAG = 'amp-mustache';
+
+/** @enum {boolean} */
+const DisallowedAdAttributes = {
+  'height': true,
+  'layout': true,
+  'width': true,
+};
+
+/** @enum {boolean} */
+const AllowedAdTypes = {
+  'custom': true,
+  'doubleclick': true,
+  'fake': true,
+};
+
+export class StoryAdConfig {
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   */
+  constructor(win, element) {
+    this.win_ = win;
+    this.element_ = element;
+  }
+
+  /**
+   * Validate and sanitize config.
+   * @return {!JsonObject}
+   */
+  getConfig() {
+    const child = this.element_.firstElementChild;
+    userAssert(
+      child && isJsonScriptTag(child),
+      `The ${TAG} should ` +
+        'be inside a <script> tag with type="application/json"'
+    );
+
+    const dirtyConfig = parseJson(child.textContent);
+
+    const requiredAttrs = {
+      class: 'i-amphtml-story-ad',
+      layout: 'fill',
+      'amp-story': '',
+    };
+
+    const adAttributes = dirtyConfig['ad-attributes'];
+    userAssert(
+      adAttributes,
+      `${TAG} Error reading config. ` +
+        'Top level JSON should have an "ad-attributes" key'
+    );
+
+    this.handleType_(adAttributes['type']);
+
+    for (const attr in adAttributes) {
+      const value = adAttributes[attr];
+      if (isObject(value)) {
+        adAttributes[attr] = JSON.stringify(value);
+      }
+      if (DisallowedAdAttributes[attr]) {
+        user().warn(TAG, 'ad-attribute "%s" is not allowed', attr);
+        delete adAttributes[attr];
+      }
+    }
+
+    return /** @type {!JsonObject} */ (Object.assign(
+      {},
+      adAttributes,
+      requiredAttrs
+    ));
+  }
+
+  /**
+   * Logic specific to each ad type.
+   * @param {string} type
+   */
+  handleType_(type) {
+    userAssert(
+      !!AllowedAdTypes[type],
+      `${TAG} "${type}" ad type is missing or not supported`
+    );
+
+    if (type === 'fake') {
+      const {id} = this.element_;
+      userAssert(
+        id && startsWith(id, 'i-amphtml-demo-'),
+        `${TAG} id must start with i-amphtml-demo- to use fake ads`
+      );
+    } else if (type === 'custom') {
+      Services.extensionsFor(this.win_)./*OK*/ installExtensionForDoc(
```

Comment:
this logic does not naturally fall into a config parsing class, especially it causes a side-effect of DOM mutation. might belong to elsewhere.